### ISSUE TITLE
No way to use WebAssembly on Chrome without 'unsafe-eval'

### ISSUE THREAD

**Issue by hackcasual** [WebAssembly/content-security-policy#7]
To give a bit of background, I am a software engineer at Tableau, on our online data visualization tool. We have been targeting WebAssembly for improving visualization interactivity. As part of our security efforts, we have been working on being able to support a restrictive CSP.

Currently chrome disables compileStreaming/instantiateStreaming with CSP on and not allowing 'unsafe-eval'. Other implementations (FF/Safari/Edge) at least allow the streaming versions of the compile/instantiate commands. The 8/8/2017 meeting notes seems to indicate that Chrome is following the original straw proposal. Chrome has implemented 'wasm-eval', but restricted it to apps/extensions.

Either treating the Response object as an origin according to the given CSP or adopting 'wasm-eval' would allow us to move forward.

**Comment by hackcasual** [comment]
I've put together a quick example here: https://s3.amazonaws.com/webassembly-chrome-csp/csp_test.html

**Comment by mikesamuel** [comment]
I think the goal was for user-agents to become consistent in how they interpret `unsafe-eval` and `wasm-eval` directives.  Issue #1 discussed how not to, and issue #2 touches on how to, but I'm not sure if consensus has been reached.

**Comment by eholk** [comment]
Right now Chrome restricts `'wasm-eval'` to web apps because it's not part of the CSP standard. I'm going to try to put together a PR this week to move along the process of making `'wasm-eval'` standard, at which point it should be pretty straightforward for Chrome to expose `'wasm-eval'` to the web.

**Comment by hackcasual** [comment]
Awesome.

**Comment by eholk** [comment]
Out of curiosity, I just tried running your test case on Safari and got:
```
[Error] TypeError: WebAssembly.instantiateStreaming is not a function. (In 'WebAssembly.instantiateStreaming(fetch('simple.wasm'), importObject)', 'WebAssembly.instantiateStreaming' is undefined)
	Global Code (csp_test.js:3)
```

Does that match what you see? Based on the table on the [Wasm CSP Proposal](https://github.com/WebAssembly/content-security-policy/blob/master/proposals/CSP.md#behavior-of-current-implementations), Safari should not allow `instantiate` or `instantiateStreaming` without `'unsafe-eval'`.

**Comment by jfbastien** [comment]
Safari currently doesn't implement streaming.

**Comment by eholk** [comment]
I just submitted the PR for `'wasm-eval'`: https://github.com/w3c/webappsec-csp/pull/293

**Comment by hackcasual** [comment]
Has there been any movement on this? Looks like the PR has been open for over 4 months now.

**Comment by binji** [comment]
The CSP proposal stalled out as @eholk has moved on to another team. However, @titzer  is starting to help out now.

**Comment by hackcasual** [comment]
Awesome, thanks for pushing this forward.

**Comment by twilco** [comment]
So just for the sake of clarity, the end goal here will be to create a CSP that allows WASM without allowing `unsafe-eval` of everything?  I'm also looking at incorporating WebAssembly into my application and ran into this issue, so was hoping `unsafe-eval` wasn't the only way out.

**Comment by titzer** [comment]
@twilco Yes, the proposed `wasm-eval` (or maybe `unsafe-wasm-eval`) would be a separate directive that enables WASM code generation independent of JavaScript's `eval`.

**Comment by titzer** [comment]
There's a PR #17 to rename the directive. However, for Chrome apps that use extension URLs, we'll likely keep the old directive for some transition time.

**Comment by uBlock-user** [comment]
So has `wasm-unsafe-eval` been implemented in chrome and in which version ?

**Comment by titzer** [comment]
Hi, we've only implemented `wasm-eval` (preliminary name) in Chrome for extension URLs. The new name (and applying to all URLs) has not yet been implemented. Given that this proposal seems to be (slowly) iterating to a decent state, I think we can begin implementation soon.

**Comment by gorhill** [comment]
> Either treating the Response object as an origin according to the given CSP

I wish to be able to use WebAssembly from my extension -- this is already the case for the Firefox version of the extension.

However, I do not plan to use `wasm-unsafe-eval` if/when it becomes available, as this goes against the spirit of **not** executing remote code.

What is the reason Chromium can't allow loading of wasm modules from same-origin Response objects? This would solve the issue of loading/executing only same-origin wasm code (i.e. the extension's packaged files) without having to resort to relaxing the default CSP ni the manifest.

**Comment by feross** [comment]
I opened a bug on the Chromium bug tracker for implementing `wasm-unsafe-eval`: https://bugs.chromium.org/p/chromium/issues/detail?id=948834&can=1&q=wasm-eval Please star it if you're interested in following updates.

**Comment by robyoder** [comment]
I've filed [Support origins and SRI-style hashes in CSP for WebAssembly streaming APIs](https://bugs.chromium.org/p/chromium/issues/detail?id=961485) to address the concerns of @gorhill and myself.

**Comment by kendraspock** [comment]
Hi there - @hackcasual - I am running into this error when attempting to embed Google Street View in a dashboard in Tableau Online. I am new to coding and am curious if this conversation, particularly what you mentioned about Tableau, has to do with this error and if it's resolvable:

![image](https://user-images.githubusercontent.com/55453483/65083739-33292b00-d95e-11e9-8503-ac9116d55938.png)

TYIA!

**Comment by hackcasual** [comment]
Hi @kendraspock , that error is a direct result of this. It's a little confusing since it doesn't actually mean Chrome refused to compile the WebAssembly module, it's reporting that if CSP were enabled, it would refuse to compile the module. It's benign.

**Comment by martinberlin** [comment]
I implemented successfully a compression protocol called Brotli here:
http://udpx.fasani.de
(It's an old fashioned Chrome app)
Even if wasm-eval seems to work I see that on many Web Assembly existing libraries they use eval() so that is a big restriction to use them for me right now. The funny thing is that I had to upload the binary WASM on a server just to give it back with the right CSP:

    header("Content-Type: application/wasm");
    header("Access-Control-Allow-Origin: *");

otherwise it won't work, but I though apps restrict scripts from remote URLs.

**Comment by FloodGames** [comment]
![image](https://user-images.githubusercontent.com/60991890/99800014-f180ab00-2b33-11eb-8444-68411a7d0c24.png)
annoyed by this error

**Comment by brycetham** [comment]
Has there been any update on this? Is the only way to use WebAssembly on websites on Chrome still to add 'unsafe-eval' to the CSP?

**Comment by fgmccabe** [comment]
There has not been much progress on this in recent years.
There is an implementation of wasm-eval for chrome but it is restricted to ChromeOS at the moment.
All that being said, it is clear that the priority of this is rising and we are beginning to pick up the dusty threads :)

**Comment by Poopooracoocoo** [comment]
Argghhh I really want this for Ruffle as a user.

**Comment by DHunte** [comment]
Just encountered the same issues with firefox, edge and chrome whilst trying to use wasm generated by blazor. Its a real stumbling block in moving from local development to production. Im quite surprised that there has'nt been more progression on this front given the push behind languages to adopt WASM compilation. Also im not sure about the adoption rate of browser updates across multiple platforms and regions etc however I am concerned because I assume that when this feature does eventually release there will be an issue where any application you're trying to rollout safely will for a long period of time only be accessible by a minority of browser users.

**Comment by hackcasual** [comment]
> Just encountered the same issues with firefox

Surprised you encountered this in firefox, as far as I know firefox doesn't gate wasm compilation.

**Comment by hackcasual** [comment]
Now there is technically a way to use WASM on Chrome without `unsafe-eval`, as long as you can load a worker with unsafe-eval. This bypass doesn't work in Safari. Depending on your security needs, you might want to do something like per agent CSP policy

https://github.com/hackcasual/ChromeWASMCSP

**Comment by DHunte** [comment]
> > Just encountered the same issues with firefox
> 
> Surprised you encountered this in firefox, as far as I know firefox doesn't gate wasm compilation.
> 

It doesent block by default, the issue is if you deploy your application on a server with any kind of csp it seems to occur, though I suppose it's the same as enabling with a local document. 

It's also unavoidable since it seems the method of using a js hash as you normally would with javascript in the csp causes it to still get blocked... And these alternative tags being discussed in the this thread when used in the csp also don't allow the execution either... In short csp equals no execution atleast in my tests (which aren't exhaustive as I mentioned).

But generally anywhere I would like to use wasm in production usually requires csps to be defined since security and reputation often are more valuable than performance gains.

**Comment by whereisaaron** [comment]
Am I correct that with Chrome 89 it is **impossible** to use CSP with WASM? For my Blazor WASM app Chrome warns if I don't add `wasm-eval` Chrome won't run the WASM code. But if it add `wasm-eval` and/or `wasm-unsafe-eval` (in addition to `unsafe-eval`) Chrome say is will both _ignore_ all those and _still_ won't run WASM because no `wasm-eval` 🤣

With `script-src 'self' 'wasm-eval' 'wasm-unsafe-eval' 'unsafe-eval' 'unsafe-inline'` the CSP is basically a no-op policy at this point yet it still can't support WASM.

I appear to be full Catch-22 screwed?

![image](https://user-images.githubusercontent.com/1109585/112313898-7d7e2b80-8cfc-11eb-8d4a-6b89ecba0cb6.png)

**Comment by fgmccabe** [comment]
Currently, wasm-eval on chrome is only enabled for chrome extensions and chrome apps.
However, we are actively working on relaxing this. However (again), since this is dependent on the standards process of two separate committees within W3C, it is going to take time.

**Comment by whereisaaron** [comment]
Thanks for the background @fgmccabe I'll stay glued for updates!

Some browsers, certainly Firefox, currently allow WASM with `unsafe-eval` which keeps CSP relevant now, and provides a smooth transition to adding a slightly less 'unsafe' source specifically for WASM. Why isn't that Chrome's approach also? This current Chrome policy just forces people to _not have any CSP at all_. That doesn't feel like a sane interim position while the committees try to commit 😀

Refs:
https://github.com/dotnet/aspnetcore/issues/31190
https://github.com/w3c/webappsec-csp/pull/293
https://github.com/WebAssembly/content-security-policy/issues/29

**Comment by whereisaaron** [comment]
This situation is not quite as it appeared (to me). If you include `unsafe-eval` in a CSP policy in report-only mode, then Chrome warns it will block WASM, as discussed above. But if you ignore the error message and deploy the CSP in enforcing mode anyway, the error disappears and Chrome lets the WASM module run anyway... So it appear `unsafe-eval` _does_ currently work for Chrome, but Chrome reports some bogus errors in report-only mode, perhaps some confusion by Chrome between web pages and Chrome extensions/apps?

![image](https://user-images.githubusercontent.com/1109585/112618948-d247b080-8e7a-11eb-883c-5ea4480ea9dd.png)

So still not the final/ideal solution @fgmccabe but the current workaround does work, just ignore the report-only error.

**Comment by whereisaaron** [comment]
Further information for misleading errors on Chrome 89: If you switch to enforcing mode with a CSP policy that includes `script-src unsafe-eval`, then the CSP 'report-only' error saying Chrome will block running WebAssembly goes away, and the WebAssembly runs anyway. However, you will still get this 'issue' in the dev tools claiming your WebAssembly _was_ blocked, and telling you to add 'unsafe-eval' (which of course you already had). Basically all Chrome errors about WebAssembly are bogus and/or misleading. Best ignored.

**TL;DR You get this 'issue' below saying your WebAssembly was blocked by CSP, when in fact your WebAssembly was not blocked.**

![image](https://user-images.githubusercontent.com/1109585/112716520-fd91d480-8f3a-11eb-858c-aea7cb58dafe.png)

**Comment by whereisaaron** [comment]
@coderanger tracked down a Chrome bug report for the errant behaviors above.👏 The situation remains as per the title of this issue: We have to add `script-src unsafe-eval` to enable WebAssembly (and that works, despite the errors displayed) until a better, wasm-specific option is available. https://bugs.chromium.org/p/chromium/issues/detail?id=915648

**Comment by dm17** [comment]
Considering it is closed, can someone point me to which comment in this thread remedies the situation described here:
```The Firefox version of uBO makes use of WebAssembly code for core filtering code paths. This is not the case with Chromium-based browsers because this would require an extra permission in the extension manifest which could cause friction when publishing the extension in the Chrome Web Store.``` --
https://github.com/gorhill/uBlock/wiki/uBlock-Origin-works-best-on-Firefox

**Comment by fgmccabe** [comment]
A small follow-up. We are 'looking into this'. Not a whole lot to report yet.

**Comment by rugk** [comment]
@fgmccabe So you were looking into this, is there any result yet?
And if you are looking into this, could you probably open this issue again to show that it is "in progress" as you say?

**Comment by rugk** [comment]
Ah or is not this fixed by https://github.com/ruffle-rs/ruffle/pull/6026?

Apparently Chrome 97 now supports a directive called `wasm-eval`.

**Comment by abishekmuthian** [comment]
@rugk wasm-eval does nothing, At least on Linux x86_64 tested with Chromium v99. Others have mentioned it earlier as well, Currently we're forced use 'unsafe-eval' if we need wasm.

**Comment by fgmccabe** [comment]
wasm-eval is/was always restricted to chrome extensions and chrome apps. For manifest v3, it will likely not be supported. 
The correct csp source keyword to use is wasm-unsafe-eval.
This has been supported in chrome since m97. A patch to fix it for manifest v3 is rolling out at the moment.

**Comment by abishekmuthian** [comment]
Thanks for clarifying, `wasm-unsafe-eval` works for my web app.

**Comment by naviln** [comment]
> wasm-eval is/was always restricted to chrome extensions and chrome apps. For manifest v3, it will likely not be supported. The correct csp source keyword to use is wasm-unsafe-eval. This has been supported in chrome since m97. A patch to fix it for manifest v3 is rolling out at the moment.

Thanks!

**Comment by ivancuric** [comment]
> I've put together a quick example here: [s3.amazonaws.com/webassembly-chrome-csp/csp_test.html](https://s3.amazonaws.com/webassembly-chrome-csp/csp_test.html)

> Thanks for clarifying, `wasm-unsafe-eval` works for my web app.

I might be missing something but switching the CSP to the following still doesn't work in either Chrome 109, Firefox 109 or Safari 16.2:
```html
        <meta http-equiv="Content-Security-Policy" content="script-src 'self' 'wasm-unsafe-eval'">
```

**Comment by fgmccabe** [comment]
Can you be more specific?
wasm-unsafe-eval is specified to only impact WebAssembly execution

**Comment by OnurGumus** [comment]
@fgmccabe Sure, I was lighthouse from dev-tools itself with 
```
script-src: 'none'
```

That was mysteriously lowering the Best Practices score to 92. Then I ran CSP with report only mode and it pointed out the lighthouse tool itself was using web assembly to do its diagnosis. 

So I changed it the following and now score is back to 100% :

```
script-src 'wasm-unsafe-eval'
```

P.S my page itself doesn't use WASM anywhere.

**Comment by fgmccabe** [comment]
Got it.

**Comment by Tectract** [comment]
Please bump. This appears to be affecting the loading of Facebook content for me. As in facebook no longer works in Chrome for me. How do I override this setting and let Facebook load content again?

**Comment by rugk** [comment]
@Tectract this is then either a problem of Facebook or your browser. As the history shows this bug has been fixed: https://github.com/WebAssembly/content-security-policy/issues/7#issuecomment-1481424334

**Comment by Tectract** [comment]
If the problem was fixed why is Facebook not loading content and I'm seeing this error in the console? 

> Chrome has implemented 'wasm-eval', but restricted it to apps/extensions.

how do I override this to allow Facebook to load content again?

**Comment by rugk** [comment]
Please ask Facebook for your problems, there is likely something wrong with your side and I don't know how you came to this issue. Many people use Facebook in Chrome and any problem would quickly surface. 
Maybe try reinstalling your browser/another OS/device etc.
